lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -3782,8 +3785,11 @@ bool SROA_Helper::LowerMemcpy(Value *V, DxilFieldAnnotation *annotation,
           hlutil::PointerStatus SrcPS(Src, size, /*bLdStOnly*/ false);
           SrcPS.analyze(typeSys, bStructElt);
           if (SrcPS.storedType != hlutil::PointerStatus::StoredType::Stored) {
-            if (ReplaceMemcpy(V, Src, MC, annotation, typeSys, DL, DT))
-              return true;
+            if (ReplaceMemcpy(V, Src, MC, annotation, typeSys, DL, DT)) {
+              if (V->user_empty())
+                return true;
+              return LowerMemcpy(V, annotation, typeSys, DL, DT, bAllowReplace);
+            }
```

Comment:
For the example you gave, I don't think a recursive call is necessary so much as storing the result instead of returning here and allowing fallthrough to the code that handles the MemcopySrcOnce case. 

As is, If there is only MemcopyDestOnce, won't it needlessly fall through to the memcpysplitter on the second invocation?

Reply:
There might be memcpy's from reads, which could be cleaned up by a recursive call. We'd have to modify `hlutil::PointerStatus PS` to remove the the memcpy from its memcpy set, otherwise the fallthrough will try to split the deleted memcpy.